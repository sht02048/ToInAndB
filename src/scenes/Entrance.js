import Bug from "../entities/enemies/Bug";
import Heavy from "../entities/enemies/Heavy";
import MODIFIER from "../constants/modifier";
import Renderer from "../graphics/Renderer";

class Entrance extends Renderer {
  #appearanceFrame = 300 * MODIFIER.FRAME;
  #heavyWidth = 50;
  #isDone = true;

  constructor() {
    super();

    this.bugList = Array.from({ length: 5 }, (_, index) => [
      new Bug({ x: this.minX + 100, y: index * -150 - 50, isLeft: true }),
      new Bug({ x: this.maxX - 100, y: index * -150 - 50, isLeft: false }),
    ]).flat();
    this.heavy = new Heavy(this.canvasWidth / 2 - this.#heavyWidth / 2, -200);

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update() {
    if (this.frame > this.#appearanceFrame) {
      this.bugList.forEach((bug) => bug.update());
      this.heavy.update();
    }

    this.frame += 1;
  }

  render() {
    if (this.frame > this.#appearanceFrame) {
      this.bugList.forEach((bug) => bug.render());
      this.heavy.render();
    }
  }

  setTarget(player) {
    this.bugList.forEach((bug) => bug.setTargetList(player));
    this.heavy.setTargetList(player);
  }

  setSceneTargetList() {
    const sceneTargetList = [this.heavy];

    this.bugList.forEach((bug) => sceneTargetList.push(bug));

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.#isDone = true;

    this.bugList.forEach((bug) => {
      if (!bug.isDestroyed && !bug.isVanished) {
        this.#isDone = false;
      }
    });

    if (!this.heavy.isDestroyed && !this.heavy.isVanished) {
      this.#isDone = false;
    }

    return this.#isDone;
  }
}

export default Entrance;
