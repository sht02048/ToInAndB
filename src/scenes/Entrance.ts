import Renderer from "../graphics/Renderer";

import Player from "../entities/Player";
import Bug from "../entities/enemies/Bug";
import Heavy from "../entities/enemies/Heavy";

class Entrance extends Renderer {
  #appearanceFrame = 300;
  #heavyWidth = 50;
  #isDone = true;

  private bugList: Bug[];
  private heavy: Heavy;
  private shouldBeDisplayed: boolean;
  private hasStarted: boolean;

  constructor() {
    super();

    this.bugList = Array.from({ length: 5 }, (_, index) => [
      new Bug({ x: this.minX + 100, y: index * -150 - 50, isLeft: true }),
      new Bug({ x: this.maxX - 100, y: index * -150 - 50, isLeft: false }),
    ]).flat();
    this.heavy = new Heavy(this.canvasWidth / 2 - this.#heavyWidth / 2, -200);

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update(): void {
    if (this.frame > this.#appearanceFrame) {
      this.bugList.forEach((bug) => bug.update());
      this.heavy.update();
    }

    this.frame += 1;
  }

  render(): void {
    if (this.frame > this.#appearanceFrame) {
      this.bugList.forEach((bug) => bug.render());
      this.heavy.render();
    }
  }

  setTarget(player: Player): void {
    this.bugList.forEach((bug) => bug.setTargetList(player));
    this.heavy.setTargetList(player);
  }

  setSceneTargetList(): (Heavy | Bug)[] {
    const sceneTargetList = [this.heavy, ...this.bugList];

    return sceneTargetList;
  }

  checkSceneStatus(): boolean {
    this.#isDone = true;

    this.bugList.forEach((bug) => {
      if (!bug.isDestroyed && !bug.isVanished) {
        this.#isDone = false;
      }
    });

    if (!this.heavy.isDestroyed && !this.heavy.isVanished) {
      this.#isDone = false;
    }

    return this.#isDone;
  }
}

export default Entrance;
