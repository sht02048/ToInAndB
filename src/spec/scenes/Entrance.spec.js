import Entrance from "../../scenes/Entrance";

describe("Entrance", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let entrance;

  beforeEach(() => {
    entrance = new Entrance();
  });

  it("should have proper functions", () => {
    expect(entrance.update).to.be.a("function");
    expect(entrance.render).to.be.a("function");
    expect(entrance.setTarget).to.be.a("function");
    expect(entrance.setSceneTargetList).to.be.a("function");
    expect(entrance.checkSceneStatus).to.be.a("function");
  });

  it("should render bug and heavy", () => {
    entrance.frame = 301;
    entrance.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should update bug and heavy", () => {
    entrance.frame = 301;
    entrance.update();
    entrance.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should check if scene is done", () => {
    entrance.bugList.forEach((bug) => (bug.isDestroyed = true));
    entrance.heavy.isDestroyed = true;

    entrance.checkSceneStatus();

    expect(entrance.checkSceneStatus()).to.equal(true);
  });
});
