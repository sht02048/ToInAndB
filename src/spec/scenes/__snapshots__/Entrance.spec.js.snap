// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Entrance > should render bug and heavy 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -50,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -50,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -350,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -350,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -500,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -500,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -650,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -650,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 125,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/heavy.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`Entrance > should update bug and heavy 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -50,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -50,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -350,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -350,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -500,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -500,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -650,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -650,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 125,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/heavy.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 220,
      "dy": 6,
      "img": <img
        src="/src/assets/images/projectile_1_enemy.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 204.3371161279701,
      "dy": -48,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 75,
      "dy": 6,
      "img": <img
        src="/src/assets/images/projectile_1_enemy.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 59.337116127970084,
      "dy": -48,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 204.3371161279701,
      "dy": -198,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 59.337116127970084,
      "dy": -198,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 204.3371161279701,
      "dy": -348,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 59.337116127970084,
      "dy": -348,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 204.3371161279701,
      "dy": -498,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 59.337116127970084,
      "dy": -498,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 204.3371161279701,
      "dy": -648,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 59.337116127970084,
      "dy": -648,
      "img": <img
        src="/src/assets/sprites/enemies/bug.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 125,
      "dy": -199,
      "img": <img
        src="/src/assets/sprites/enemies/heavy.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
